# Pandas Data Frames

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/rhennig/EMA6938/blob/main/Notebooks/5.Pandas_DataFrames.ipynb)

(Based https://www.w3schools.com/python/pandas/ and https://realpython.com/python-data-cleaning-numpy-pandas/)

This notebook introduces the Panda Data Frames library. We will learn how to easily load and manipulate data, from selecting or replacing columns and indices to reshaping the data.

### First, load needed Python packages

In [1]:
# Import numpy and pandas dataframes
import numpy as np
import pandas as pd

# Set pandas view options
pd.set_option('display.width', 800)
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

## 1. Pandas Data Structures

Pandas provides two types of data structures: **Series** and **DataFrames**:

- A **Series** is a one dimensional data structure (“a one dimensional ndarray”) that can store values — and for every value it holds a unique index.

- A **DataFrame** is a two (or more) dimensional data structure – basically a table with rows and columns. The columns have names and the rows have indexes.

<img src="https://github.com/rhennig/EMA6938/blob/main/Notebooks/Figures/Pandas.png?raw=1" alt="Confusion Matrix" align="center" style="width:500px; float:center"/>

In [2]:
# Create a Pandas DataFrame

# Using a dict of ndarray/lists
data = {
  'Element': ['Titanium', 'Vanadium', 'Manganese',
              'Chromium', 'Iron', 'Cobalt', 'Nickel'],
  'Z': [22, 23, 24,
        25, 26, 27, 28],
  'Magnetism':['Paramagnetic', 'Paramagnetic', 'Complex',
               'Antiferromagnetic', 'Ferromagnetic', 'Ferromagnetic', 'Ferromagnetic']
}

# Load data into a DataFrame object:
df = pd.DataFrame(data)

print(df)

     Element   Z          Magnetism
0   Titanium  22       Paramagnetic
1   Vanadium  23       Paramagnetic
2  Manganese  24            Complex
3   Chromium  25  Antiferromagnetic
4       Iron  26      Ferromagnetic
5     Cobalt  27      Ferromagnetic
6     Nickel  28      Ferromagnetic


In [3]:
# Alternatively, create Pandas DataFrame from lists of lists
data = [['Titanium', 22, 'Paramagnetic'],
        ['Vanadium', 23, 'Paramagnetic'],
        ['Manganese', 24, 'Complex'],
        ['Chromium', 25, 'Antiferromagnetic'],
        ['Iron', 26, 'Ferromagnetic'],
        ['Cobalt', 27, 'Ferromagnetic'],
        ['Nickel', 28, 'Ferromagnetic']]

# Load data into a DataFrame object:
df = pd.DataFrame(data, columns = ['Element', 'Z', 'Magnetism'])

print(df)

     Element   Z          Magnetism
0   Titanium  22       Paramagnetic
1   Vanadium  23       Paramagnetic
2  Manganese  24            Complex
3   Chromium  25  Antiferromagnetic
4       Iron  26      Ferromagnetic
5     Cobalt  27      Ferromagnetic
6     Nickel  28      Ferromagnetic


### Locate Row

As you can see from the result above, the DataFrame is like a table with rows and columns.

Pandas use the loc attribute to return one or more specified row(s).

In [4]:
# Refer to the row index:
print(df.loc[0])

Element          Titanium
Z                      22
Magnetism    Paramagnetic
Name: 0, dtype: object


Note that this example returns a Pandas Series.

In [5]:
# Use a list of indexes:
df.loc[[0, 1]]

Element   Z     Magnetism
0  Titanium  22  Paramagnetic
1  Vanadium  23  Paramagnetic

When using [], the result is a Pandas DataFrame.

In [6]:
# Refer to the column name:
df.loc[:, 'Magnetism']

0         Paramagnetic
1         Paramagnetic
2              Complex
3    Antiferromagnetic
4        Ferromagnetic
5        Ferromagnetic
6        Ferromagnetic
Name: Magnetism, dtype: object

In [7]:
# Conditional that returns a boolean Series
df.loc[df['Magnetism'] == 'Ferromagnetic']

Element   Z      Magnetism
4    Iron  26  Ferromagnetic
5  Cobalt  27  Ferromagnetic
6  Nickel  28  Ferromagnetic

### Named Indices

With the index argument, you can name your own indexes.

In [8]:
# Give each row a name
df = pd.DataFrame(data, index = ['Ti', 'V', 'Mn', 'Cr', 'Fe', 'Co', 'Ni'])

print('DataFrame:\n', df)

# Refer to the named index:
print('\nRefer to names index:')
print(df.loc['Mn'])

DataFrame:
             0   1                  2
Ti   Titanium  22       Paramagnetic
V    Vanadium  23       Paramagnetic
Mn  Manganese  24            Complex
Cr   Chromium  25  Antiferromagnetic
Fe       Iron  26      Ferromagnetic
Co     Cobalt  27      Ferromagnetic
Ni     Nickel  28      Ferromagnetic

Refer to names index:
0    Manganese
1           24
2      Complex
Name: Mn, dtype: object


## 2. Working with CSV Files

A simple and widely used method to store big data sets is to use CSV (comma separated-values) files. CSV files contains plain text and ican be read by most software packages, including Pandas.

In our examples we will be using a CSV file called cleavage_data.csv. This file contains the energy the cleave a number of materials.

In [9]:
# Read CSV file into pandas dataframe if the file is locally available
# df = pd.read_csv('Datasets/cleavage_data.csv')

# Otherwise, we can read this file directly from a Zip file on MaterialsCloud.org
from io import BytesIO
from zipfile import ZipFile
from urllib.request import urlopen

url = urlopen("https://archive.materialscloud.org/record/file?filename=theoreticalCleavedSubstrates.zip&record_id=948")

#Download Zipfile and create pandas DataFrame
zipfile = ZipFile(BytesIO(url.read()))
data = pd.read_csv(zipfile.open('database/cleavage_data.csv'))

# First five data entries
data.head()

Substrate Index Materials Project ID  Initial formation energy  Final formation energy  Initial area  Final area  Work of Cleavage  Surface Energy
0              291           mp-1215430                  0.805967                0.709898     15.672351   15.672351          0.205704        0.090592
1             3538           mp-1228894                  1.895515                1.883755      8.700654    8.700654          0.871436        0.433015
2             1060            mp-754655                  2.186383                1.585846     13.489288   13.489288          0.810414        0.293908
3             2841            mp-981544                  0.491037                0.394102     12.135709   12.135709          0.121386        0.048712
4             2031            mp-999502                  1.359986                1.262626     12.435465   12.435465          0.437454        0.203069

In [10]:
# Last five data entries
data.tail()

Substrate Index Materials Project ID  Initial formation energy  Final formation energy  Initial area  Final area  Work of Cleavage  Surface Energy
4609             1544           mp-1068235                  0.613948                0.601773     20.464395   20.464395          0.150004        0.073515
4610              368             mp-31380                  0.482304                0.461286     17.680294   17.680294          0.109117        0.052181
4611             1107           mp-1215529                  0.191980                0.188179      5.948710    5.948710          0.129090        0.063267
4612              898            mp-568993                  0.542114                0.486801     16.941106   16.941106          0.159999        0.071837
4613             2565           mp-1224571                  1.195508                0.593128     11.280699   11.280699          0.423913        0.105158

In [11]:
# Print information about the data
data.describe()

Substrate Index  Initial formation energy  Final formation energy  Initial area   Final area  Work of Cleavage  Surface Energy
count      4614.000000               4614.000000             4614.000000   4614.000000  4614.000000       4614.000000     4614.000000
mean       2342.658431                  0.777245                0.677913     17.319837    17.319837          0.243117        0.105894
std        1358.743406                  0.633077                0.598184     10.523085    10.523085          0.253263        0.118708
min           1.000000                  0.004851                0.004851      5.387589     5.387589          0.000925        0.000463
25%        1163.250000                  0.414807                0.340569     12.450114    12.450114          0.105097        0.044686
50%        2337.500000                  0.643034                0.551580     15.382083    15.382083          0.181605        0.079720
75%        3521.750000                  0.971106                0.866148     17.927848    17.927848          0.304334        0.134306
max        4692.000000                  9.360227                9.360227    131.170497   131.170497          7.147410        3.573705

In [12]:
# Information about the data
print(data.info()) 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4614 entries, 0 to 4613
Data columns (total 8 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Substrate Index           4614 non-null   int64  
 1   Materials Project ID      4614 non-null   object 
 2   Initial formation energy  4614 non-null   float64
 3   Final formation energy    4614 non-null   float64
 4   Initial area              4614 non-null   float64
 5   Final area                4614 non-null   float64
 6   Work of Cleavage          4614 non-null   float64
 7   Surface Energy            4614 non-null   float64
dtypes: float64(6), int64(1), object(1)
memory usage: 288.5+ KB
None


The result tells us there are 4614 rows and 8 columns and the name of each column, with the data type.

We also see how many null entries (no values) each column contains. This dataset contains no empty entries.

## 3. Cleaning Data

Data scientists spend a large amount of their time cleaning datasets. Often these initial steps of obtaining and cleaning data constitute about 80% of the work in a machine learning project.

Therefore, we need to learn to deal with messy data, whether that means missing values, inconsistent formatting, malformed records, or nonsensical outliers.

Here, we will use the Pandas and NumPy libraries to clean data.

We’ll cover the following:
- Dropping unnecessary columns in a DataFrame
- Changing the index of a DataFrame

In [13]:
import pandas as pd
import numpy as np

### Dropping Columns in a DataFrame

Often, not all the categories of data in a dataset are useful for our analysis. For example, a dataset may contain materials information (composition, crystal structure, thermodynamics data, mechanical, electronic, and magnetic properties) but we may want to focus on analyzing the bulk modulus.

In this case, the electronic, and magnetic properties are not important. Retaining these unneeded categories will take up unnecessary space and potentially also bog down runtime.

Pandas provides a handy way of removing unwanted columns or rows from a DataFrame with the drop() function. Let’s look at a simple example where we drop a number of columns from a DataFrame.

First, let’s use the DataFrame from the CSV file ‘cleavage_data.csv’. In the examples below, we pass a relative path to pd.read_csv, meaning that all of the datasets are in a folder named Datasets in our current working directory:

In [14]:
# First five data entries
data.head()

Substrate Index Materials Project ID  Initial formation energy  Final formation energy  Initial area  Final area  Work of Cleavage  Surface Energy
0              291           mp-1215430                  0.805967                0.709898     15.672351   15.672351          0.205704        0.090592
1             3538           mp-1228894                  1.895515                1.883755      8.700654    8.700654          0.871436        0.433015
2             1060            mp-754655                  2.186383                1.585846     13.489288   13.489288          0.810414        0.293908
3             2841            mp-981544                  0.491037                0.394102     12.135709   12.135709          0.121386        0.048712
4             2031            mp-999502                  1.359986                1.262626     12.435465   12.435465          0.437454        0.203069

When we look at the first five entries using the head() method, we can see that a handful of columns provide ancillary information that may not be of interest if we want to select a substrate materials for a synthesis experiment: `Initial formation energy`, `Final formation energy`, `Initial area`.

We can drop these columns.

In [15]:
# Define a list of column names to drop.
to_drop = ['Initial formation energy', 'Final formation energy', 'Initial area']

# Tell pandas to drop these columns directly in the dataset (inplace = True)
data.drop(columns=to_drop, inplace=True)

# First five entries of cleaned DataFrame
data.head()

Substrate Index Materials Project ID  Final area  Work of Cleavage  Surface Energy
0              291           mp-1215430   15.672351          0.205704        0.090592
1             3538           mp-1228894    8.700654          0.871436        0.433015
2             1060            mp-754655   13.489288          0.810414        0.293908
3             2841            mp-981544   12.135709          0.121386        0.048712
4             2031            mp-999502   12.435465          0.437454        0.203069

Alternatively, if we know which columns we need, we could pass the names of the columns as a list to the `usecols` argument of `pd.read_csv`.

### Changing the Index of a DataFrame

A Pandas Index extends the functionality of NumPy arrays to allow for more versatile slicing and labeling. In many cases, it is helpful to use a uniquely valued identifying field of the data as its index.

For example, in the cleavage dataset, we may want to use the `Substrate Index` as a unique identifier.

In [16]:
# Check that the Substrate Index is a unique identifier
data['Substrate Index'].is_unique

True

In [17]:
# Let’s replace the existing index with this column using set_index
data.set_index('Substrate Index', inplace=True)
data.head()

Materials Project ID  Final area  Work of Cleavage  Surface Energy
Substrate Index                                                                   
291                       mp-1215430   15.672351          0.205704        0.090592
3538                      mp-1228894    8.700654          0.871436        0.433015
1060                       mp-754655   13.489288          0.810414        0.293908
2841                       mp-981544   12.135709          0.121386        0.048712
2031                       mp-999502   12.435465          0.437454        0.203069

There are many other ways that Pandas can help us clean data, such as:
- Dealing with empty cells
- Using .str() methods to clean columns
- Using the DataFrame.applymap() function to clean the entire dataset, element-wise
- Renaming columns to a more recognizable set of labels
- Skipping unnecessary rows in a CSV file
- Remove duplicate entries

You can find information about these methods in the following two tutorials:
- https://www.w3schools.com/python/pandas/pandas_cleaning.asp
- https://realpython.com/python-data-cleaning-numpy-pandas/